// @babel/cli/lib/babel/watcher.js
"use strict";

function _toConsumableArray(r) { return _arrayWithoutHoles(r) || _iterableToArray(r) || _unsupportedIterableToArray(r) || _nonIterableSpread(); }
function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _iterableToArray(r) { if ("undefined" != typeof Symbol && null != r[Symbol.iterator] || null != r["@@iterator"]) return Array.from(r); }
function _arrayWithoutHoles(r) { if (Array.isArray(r)) return _arrayLikeToArray(r); }
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t["return"] || t["return"](); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.enable = enable;
exports.onFilesChange = onFilesChange;
exports.startWatcher = startWatcher;
exports.updateExternalDependencies = updateExternalDependencies;
exports.watch = watch;
function _path() {
  var data = require("path");
  _path = function _path() {
    return data;
  };
  return data;
}
var fileToDeps = new Map();
var depToFiles = new Map();
var isWatchMode = false;
var watcher;
var watchQueue = new Set();
var hasStarted = false;
function enable(_ref) {
  var enableGlobbing = _ref.enableGlobbing;
  isWatchMode = true;
  var _requireChokidar = requireChokidar(),
    FSWatcher = _requireChokidar.FSWatcher;
  var options = {
    disableGlobbing: !enableGlobbing,
    persistent: true,
    ignoreInitial: true,
    awaitWriteFinish: {
      stabilityThreshold: 50,
      pollInterval: 10
    }
  };
  watcher = new FSWatcher(options);
  watcher.on("unlink", unwatchFile);
}
function startWatcher() {
  hasStarted = true;
  var _iterator = _createForOfIteratorHelper(watchQueue),
    _step;
  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var dep = _step.value;
      watcher.add(dep);
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }
  watchQueue.clear();
  watcher.on("ready", function () {
    console.log("The watcher is ready.");
  });
}
function watch(filename) {
  if (!isWatchMode) {
    throw new Error("Internal Babel error: .watch called when not in watch mode.");
  }
  if (!hasStarted) {
    watchQueue.add(_path().resolve(filename));
  } else {
    watcher.add(_path().resolve(filename));
  }
}
function onFilesChange(callback) {
  if (!isWatchMode) {
    throw new Error("Internal Babel error: .onFilesChange called when not in watch mode.");
  }
  watcher.on("all", function (event, filename) {
    var _depToFiles$get;
    if (event !== "change" && event !== "add") return;
    var absoluteFile = _path().resolve(filename);
    callback([absoluteFile].concat(_toConsumableArray((_depToFiles$get = depToFiles.get(absoluteFile)) != null ? _depToFiles$get : [])), event, absoluteFile);
  });
}
function updateExternalDependencies(filename, dependencies) {
  if (!isWatchMode) return;
  var absFilename = _path().resolve(filename);
  var absDependencies = new Set(Array.from(dependencies, function (dep) {
    return _path().resolve(dep);
  }));
  var deps = fileToDeps.get(absFilename);
  if (deps) {
    var _iterator2 = _createForOfIteratorHelper(deps),
      _step2;
    try {
      for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
        var dep = _step2.value;
        if (!absDependencies.has(dep)) {
          removeFileDependency(absFilename, dep);
        }
      }
    } catch (err) {
      _iterator2.e(err);
    } finally {
      _iterator2.f();
    }
  }
  var _iterator3 = _createForOfIteratorHelper(absDependencies),
    _step3;
  try {
    for (_iterator3.s(); !(_step3 = _iterator3.n()).done;) {
      var _dep = _step3.value;
      var _deps = depToFiles.get(_dep);
      if (!_deps) {
        depToFiles.set(_dep, _deps = new Set());
        if (!hasStarted) {
          watchQueue.add(_dep);
        } else {
          watcher.add(_dep);
        }
      }
      _deps.add(absFilename);
    }
  } catch (err) {
    _iterator3.e(err);
  } finally {
    _iterator3.f();
  }
  fileToDeps.set(absFilename, absDependencies);
}
function removeFileDependency(filename, dep) {
  var deps = depToFiles.get(dep);
  deps["delete"](filename);
  if (deps.size === 0) {
    depToFiles["delete"](dep);
    if (!hasStarted) {
      watchQueue["delete"](dep);
    } else {
      watcher.unwatch(dep);
    }
  }
}
function unwatchFile(filename) {
  var deps = fileToDeps.get(filename);
  if (!deps) return;
  var _iterator4 = _createForOfIteratorHelper(deps),
    _step4;
  try {
    for (_iterator4.s(); !(_step4 = _iterator4.n()).done;) {
      var dep = _step4.value;
      removeFileDependency(filename, dep);
    }
  } catch (err) {
    _iterator4.e(err);
  } finally {
    _iterator4.f();
  }
  fileToDeps["delete"](filename);
}
function requireChokidar() {
  try {
    return parseInt(process.versions.node) >= 8 ? require("chokidar") : require("@nicolo-ribaudo/chokidar-2");
  } catch (err) {
    console.error("The optional dependency chokidar failed to install and is required for " + "--watch. Chokidar is likely not supported on your platform.");
    throw err;
  }
}
