// @babel/helper-builder-binary-assignment-operator-visitor/lib/explode-assignable-expression.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = explode;
var _t = require("@babel/types");
var assignmentExpression = _t.assignmentExpression,
  cloneNode = _t.cloneNode,
  isIdentifier = _t.isIdentifier,
  isLiteral = _t.isLiteral,
  isMemberExpression = _t.isMemberExpression,
  isPrivateName = _t.isPrivateName,
  isPureish = _t.isPureish,
  isSuper = _t.isSuper,
  memberExpression = _t.memberExpression,
  toComputedKey = _t.toComputedKey;
function getObjRef(node, nodes, scope) {
  var ref;
  if (isIdentifier(node)) {
    if (scope.hasBinding(node.name)) {
      return node;
    } else {
      ref = node;
    }
  } else if (isMemberExpression(node)) {
    ref = node.object;
    if (isSuper(ref) || isIdentifier(ref) && scope.hasBinding(ref.name)) {
      return ref;
    }
  } else {
    throw new Error("We can't explode this node type ".concat(node["type"]));
  }
  var temp = scope.generateUidIdentifierBasedOnNode(ref);
  scope.push({
    id: temp
  });
  nodes.push(assignmentExpression("=", cloneNode(temp), cloneNode(ref)));
  return temp;
}
function getPropRef(node, nodes, scope) {
  var prop = node.property;
  if (isPrivateName(prop)) {
    throw new Error("We can't generate property ref for private name, please install `@babel/plugin-transform-class-properties`");
  }
  var key = toComputedKey(node, prop);
  if (isLiteral(key) && isPureish(key)) return key;
  var temp = scope.generateUidIdentifierBasedOnNode(prop);
  scope.push({
    id: temp
  });
  nodes.push(assignmentExpression("=", cloneNode(temp), cloneNode(prop)));
  return temp;
}
function explode(node, nodes, scope) {
  var obj = getObjRef(node, nodes, scope);
  var ref, uid;
  if (isIdentifier(node)) {
    ref = cloneNode(node);
    uid = obj;
  } else {
    var prop = getPropRef(node, nodes, scope);
    var computed = node.computed || isLiteral(prop);
    uid = memberExpression(cloneNode(obj), cloneNode(prop), computed);
    ref = memberExpression(cloneNode(obj), cloneNode(prop), computed);
  }
  return {
    uid: uid,
    ref: ref
  };
}
